# It is a regression problem.


## We are doing the machine learning on house price dataset. 

<table align = "center" width="100%">
<tr>
</td>
</td>
</td>
<td>
<div align="center">
<font color="#21618c" size="8px">
<b>  Regression modeling on House Price Data </b>
</font>
</div>
</td>
</tr>
</table>

# Problem Statement

This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.

In [106]:
# Import Pandas
# To work with DataFrames we use Pandas
# We use pandas for data manipulation, data cleaning and data analysis
# Pandas is utilised to format the data structure. It means it can convert any data structure into a Pandas DataFrame
import pandas as pd

# Import Numpy
# To find Statistical Summaries we use Numpy
import numpy as np

# Import subpackage of matplotlib 
import matplotlib.pyplot as plt

# Import Seaborn
import seaborn as sns

# To Supress Warnings
from warnings import filterwarnings
filterwarnings('ignore')



### Load the dataset

In [107]:
# Load the data and name it as df_house_price
df_house_price = pd.read_csv('house_price.csv')

#display the first five observations using head()
df_house_price.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

## Data Overview 

In [108]:
# Name the Coloumns
df_house_price.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

- `price` : Price of the House. 
- `resid_area` : Resedential Area where my house is located
- `air_qual` : Air Quality near the house
- ``
- ``
  

In [109]:
# Basic Info of the Dataset
df_house_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**
 - We have 13 Numerical columns & 3 Categorical Columns
 - We have null values in the data
 - We have total 16 columns in the data
  

In [110]:
# Shape of the data
df_house_price.shape

(506, 16)

**Interpretations** 
- We have total 16 columns 
- We have 506 Observations

**Interpretation**
 - We have outliers in the data , most of the outliers are in age coloumn

## Data Preprocessing

### A . Data Cleaning

In [112]:
# Show the First 5 Observations
df_house_price.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [113]:
# Rename the Columns
# Step 1 : Show the Column Names
df_house_price.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [114]:
# Step 2 : Allocate ew column names to Old Column Names
# Make the Dictionary {'Old Column Name' : 'New Column Name'}
d1 = {'price' : 'Price',
      'resid_area' : 'Residential Area',
      'air_qual' : 'Air Quality',
      'room_num' : 'Room Numbers',
      'age' : 'Age',
      'distance' : 'Distance',
      'teachers' : 'Teachers',
      'poor_prop' : 'Proportion of Poor Population',
      'airport' : 'Airport',
      'n_hos_beds' : 'Number of Hospital Beds',
      'n_hot_rooms' : 'Number of Hotel Rooms',
      'waterbody' : 'Waterbody',
      'rainfall' : 'Rainfall',
      'bus_ter' : 'Bus Terminal',
      'parks' : 'Parks',
      'Sold' : 'Sold'}


In [115]:
df_house_price = df_house_price.rename(columns = d1)

In [116]:
# Authenticate the Results
df_house_price.columns

Index(['Price', 'Residential Area', 'Air Quality', 'Room Numbers', 'Age',
       'Distance', 'Teachers', 'Proportion of Poor Population', 'Airport',
       'Number of Hospital Beds', 'Number of Hotel Rooms', 'Waterbody',
       'Rainfall', 'Bus Terminal', 'Parks', 'Sold'],
      dtype='object')

### B. Null Value Handling

```
Laws that we need to follow while processing the null values (Specially for Manufacturing)
 - When we have less than 15% data missing -> Remove the observation
 - Between 20% to 70% data missing -> Impute the null values
 - Greater than 75% Values missing -> Remove the Column
```

In [117]:
# Find the Valid Null Values
df_house_price.isnull().sum()

Price                              0
Residential Area                   0
Air Quality                        0
Room Numbers                       0
Age                                0
Distance                           0
Teachers                           0
Proportion of Poor Population      0
Airport                            0
Number of Hospital Beds            8
Number of Hotel Rooms              0
Waterbody                        155
Rainfall                           0
Bus Terminal                       0
Parks                              0
Sold                               0
dtype: int64

In [118]:
# Percentage of Missing Values
round(df_house_price.isnull().sum() / df_house_price.shape[0] * 100),3

(Price                             0.0
 Residential Area                  0.0
 Air Quality                       0.0
 Room Numbers                      0.0
 Age                               0.0
 Distance                          0.0
 Teachers                          0.0
 Proportion of Poor Population     0.0
 Airport                           0.0
 Number of Hospital Beds           2.0
 Number of Hotel Rooms             0.0
 Waterbody                        31.0
 Rainfall                          0.0
 Bus Terminal                      0.0
 Parks                             0.0
 Sold                              0.0
 dtype: float64,
 3)

In [119]:
# Find the Nature of the Column 
df_house_price.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Price                          506 non-null    float64
 1   Residential Area               506 non-null    float64
 2   Air Quality                    506 non-null    float64
 3   Room Numbers                   506 non-null    float64
 4   Age                            506 non-null    float64
 5   Distance                       506 non-null    float64
 6   Teachers                       506 non-null    float64
 7   Proportion of Poor Population  506 non-null    float64
 8   Airport                        506 non-null    object 
 9   Number of Hospital Beds        498 non-null    float64
 10  Number of Hotel Rooms          506 non-null    float64
 11  Waterbody                      351 non-null    object 
 12  Rainfall                       506 non-null    int

In [120]:
# This is a Categorical Column so we can replace the null with the mode
df_house_price.Waterbody.mode()


0    River
Name: Waterbody, dtype: object

In [121]:
# Replace the Null Values
df_house_price.Waterbody =  df_house_price.Waterbody.fillna('River')

In [122]:
round(df_house_price.isnull().sum() / df_house_price.shape[0] * 100),3

(Price                            0.0
 Residential Area                 0.0
 Air Quality                      0.0
 Room Numbers                     0.0
 Age                              0.0
 Distance                         0.0
 Teachers                         0.0
 Proportion of Poor Population    0.0
 Airport                          0.0
 Number of Hospital Beds          2.0
 Number of Hotel Rooms            0.0
 Waterbody                        0.0
 Rainfall                         0.0
 Bus Terminal                     0.0
 Parks                            0.0
 Sold                             0.0
 dtype: float64,
 3)

In [123]:
# In the Column 'Number of Hospital Beds' so we can drop the null values
df_house_price.dropna(inplace=True)

In [124]:
round(df_house_price.isnull().sum() / df_house_price.shape[0] * 100),3

(Price                            0.0
 Residential Area                 0.0
 Air Quality                      0.0
 Room Numbers                     0.0
 Age                              0.0
 Distance                         0.0
 Teachers                         0.0
 Proportion of Poor Population    0.0
 Airport                          0.0
 Number of Hospital Beds          0.0
 Number of Hotel Rooms            0.0
 Waterbody                        0.0
 Rainfall                         0.0
 Bus Terminal                     0.0
 Parks                            0.0
 Sold                             0.0
 dtype: float64,
 3)

In [125]:
## Prepare Data for Machine Learning 
# We Know that the Machine Learning Algorithms are Mathematical Expressions so we need to provide the data to the algorithm in specific format
# Encoding 

# Step 1 Divide the data based on thier data type 
df_cat = df_house_price.select_dtypes(include = ['object'])


df_cat.columns

Index(['Airport', 'Waterbody', 'Bus Terminal'], dtype='object')

In [126]:
# Use Dummy Encoding
dummy_var = pd.get_dummies(data = df_cat, drop_first = True)
dummy_var = dummy_var.reset_index()

In [127]:
dummy_var2 = pd.get_dummies(df_cat) 

In [128]:
dummy_var2.head(2)

Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0       False         True           False                     False   
1        True        False            True                     False   

   Waterbody_River  Bus Terminal_YES  
0             True              True  
1            False              True

In [129]:
df_num = df_house_price.select_dtypes(include = [np.number]) 
df_num.head()

Price  Residential Area  Air Quality  Room Numbers   Age  Distance  \
0   24.0             32.31        0.538         6.575  65.2      4.35   
1   21.6             37.07        0.469         6.421  78.9      4.99   
2   34.7             37.07        0.469         7.185  61.1      5.03   
3   33.4             32.18        0.458         6.998  45.8      6.21   
4   36.2             32.18        0.458         7.147  54.2      6.16   

   Teachers  Proportion of Poor Population  Number of Hospital Beds  \
0      24.7                           4.98                    5.480   
1      22.2                           9.14                    7.332   
2      22.2                           4.03                    7.394   
3      21.3                           2.94                    9.268   
4      21.3                           5.33                    8.824   

   Number of Hotel Rooms  Rainfall     Parks  Sold  
0                11.1920        23  0.049347     0  
1                12.1728        42  0.046146     1  
2               101.1200        38  0.045764     0  
3                11.2672        45  0.047151     0  
4                11.2896        55  0.039474     0

In [130]:
from sklearn.preprocessing import StandardScaler


In [131]:
df_target = df_house_price['Price']
df_num = df_num.drop('Price',axis = 1)

In [132]:
# Initiate the StandardScaler 
ss = StandardScaler()

In [133]:
# Transform the data
Scale = ss.fit_transform(df_num)
Scale

array([[-1.28244435, -0.13709359,  0.40897478, ..., -1.30206786,
        -0.47313205, -0.91152748],
       [-0.59005882, -0.73319492,  0.18997385, ...,  0.21519397,
        -0.77522463,  1.09705963],
       [-0.59005882, -0.73319492,  1.276446  , ..., -0.10422957,
        -0.81123666, -0.91152748],
       ...,
       [ 0.11687263,  0.16527665,  0.97923045, ..., -0.66322077,
         0.30293265,  1.09705963],
       [ 0.11687263,  0.16527665,  0.72041117, ...,  0.6144734 ,
         0.59750894,  1.09705963],
       [ 0.11687263,  0.16527665, -0.36606097, ...,  0.45476163,
         0.56369225,  1.09705963]])

In [134]:
df_scale = pd.DataFrame(Scale, columns = df_num.columns)
df_scale = df_scale.reset_index()
df_scale.head()

index  Residential Area  Air Quality  Room Numbers       Age  Distance  \
0      0         -1.282444    -0.137094      0.408975 -0.115644  0.172845   
1      1         -0.590059    -0.733195      0.189974  0.369741  0.476100   
2      2         -0.590059    -0.733195      1.276446 -0.260905  0.495053   
3      3         -1.301354    -0.828226      1.010516 -0.802977  1.054179   
4      4         -1.301354    -0.828226      1.222407 -0.505369  1.030487   

   Teachers  Proportion of Poor Population  Number of Hospital Beds  \
0  1.462167                      -1.066368                -1.640299   
1  0.304963                      -0.484136                -0.384875   
2  0.304963                      -1.199330                -0.342847   
3 -0.111631                      -1.351886                 0.927490   
4 -0.111631                      -1.017382                 0.626514   

   Number of Hotel Rooms  Rainfall     Parks      Sold  
0              -0.352111 -1.302068 -0.473132 -0.911527  
1              -0.166019  0.215194 -0.775225  1.097060  
2              16.710340 -0.104230 -0.811237 -0.911527  
3              -0.337843  0.454762 -0.680402 -0.911527  
4              -0.333593  1.253320 -1.404723 -0.911527

In [135]:
# Make a DataFrame for independable variables
df_independent = pd.concat([df_scale, dummy_var], axis = 1)
df_independent.head()

index  Residential Area  Air Quality  Room Numbers       Age  Distance  \
0      0         -1.282444    -0.137094      0.408975 -0.115644  0.172845   
1      1         -0.590059    -0.733195      0.189974  0.369741  0.476100   
2      2         -0.590059    -0.733195      1.276446 -0.260905  0.495053   
3      3         -1.301354    -0.828226      1.010516 -0.802977  1.054179   
4      4         -1.301354    -0.828226      1.222407 -0.505369  1.030487   

   Teachers  Proportion of Poor Population  Number of Hospital Beds  \
0  1.462167                      -1.066368                -1.640299   
1  0.304963                      -0.484136                -0.384875   
2  0.304963                      -1.199330                -0.342847   
3 -0.111631                      -1.351886                 0.927490   
4 -0.111631                      -1.017382                 0.626514   

   Number of Hotel Rooms  Rainfall     Parks      Sold  index  Airport_YES  \
0              -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1              -0.166019  0.215194 -0.775225  1.097060      1        False   
2              16.710340 -0.104230 -0.811237 -0.911527      2        False   
3              -0.337843  0.454762 -0.680402 -0.911527      3         True   
4              -0.333593  1.253320 -1.404723 -0.911527      4        False   

   Waterbody_Lake and River  Waterbody_River  
0                     False             True  
1                     False            False  
2                     False             True  
3                     False            False  
4                     False            False

In [136]:
# Split the data for training and testing
import statsmodels
import statsmodels.api as sm

df_independent = sm.add_constant(df_independent) 
df_independent.head()



const  index  Residential Area  Air Quality  Room Numbers       Age  \
0    1.0      0         -1.282444    -0.137094      0.408975 -0.115644   
1    1.0      1         -0.590059    -0.733195      0.189974  0.369741   
2    1.0      2         -0.590059    -0.733195      1.276446 -0.260905   
3    1.0      3         -1.301354    -0.828226      1.010516 -0.802977   
4    1.0      4         -1.301354    -0.828226      1.222407 -0.505369   

   Distance  Teachers  Proportion of Poor Population  Number of Hospital Beds  \
0  0.172845  1.462167                      -1.066368                -1.640299   
1  0.476100  0.304963                      -0.484136                -0.384875   
2  0.495053  0.304963                      -1.199330                -0.342847   
3  1.054179 -0.111631                      -1.351886                 0.927490   
4  1.030487 -0.111631                      -1.017382                 0.626514   

   Number of Hotel Rooms  Rainfall     Parks      Sold  index  Airport_YES  \
0              -0.352111 -1.302068 -0.473132 -0.911527      0         True   
1              -0.166019  0.215194 -0.775225  1.097060      1        False   
2              16.710340 -0.104230 -0.811237 -0.911527      2        False   
3              -0.337843  0.454762 -0.680402 -0.911527      3         True   
4              -0.333593  1.253320 -1.404723 -0.911527      4        False   

   Waterbody_Lake and River  Waterbody_River  
0                     False             True  
1                     False            False  
2                     False             True  
3                     False            False  
4                     False            False

In [137]:
from sklearn.model_selection import train_test_split

In [138]:
# Testing
X_train, X_test, y_train, y_test = train_test_split(df_independent, df_target, random_state = 10, test_size = 0.2)

In [139]:
# print dimensios of train set
print('X_train', X_train.shape)
print('y_train', y_train.shape)

# print dimensions of test set
print('X_test', X_test.shape)
print('y_test', y_test.shape)

X_train (398, 18)
y_train (398,)
X_test (100, 18)
y_test (100,)


In [140]:
X_train.columns

Index(['const', 'index', 'Residential Area', 'Air Quality', 'Room Numbers',
       'Age', 'Distance', 'Teachers', 'Proportion of Poor Population',
       'Number of Hospital Beds', 'Number of Hotel Rooms', 'Rainfall', 'Parks',
       'Sold', 'index', 'Airport_YES', 'Waterbody_Lake and River',
       'Waterbody_River'],
      dtype='object')

## Linear Regression

In [141]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [142]:
# Create a linear regression model
model = LinearRegression()

#Fit the model to the training data
model.fit(X_train, y_train)

# Predict on the testing data
predictions = model.predict(X_test)

# Calculate mean squared error
mse = mean_squared_error(y_test,predictions)
print("Mean Squared Error : " , mse)

Mean Squared Error :  15.498182384260279


In [144]:
from sklearn.metrics import mean_absolute_error


# Calculate mean absolute error
mae = mean_absolute_error(y_test,predictions)
print("Mean Absolute Error : " , mae)

Mean Absolute Error :  3.13314291328658


In [145]:
from sklearn.metrics import root_mean_squared_error


# Calculate mean absolute error
rmse = root_mean_squared_error(y_test,predictions)
print("Mean Absolute Error : " , rmse)

Mean Absolute Error :  3.936773092808408


# Ridge Model

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error


